staging/src/k8s.io/client-go/tools/portforward/websocket-connection.go

```diff
@@ -0,0 +1,324 @@
+	}
+	reader, writer := io.Pipe()
+	s := &stream{
+		headers:       headers,
+		readPipe:      reader,
+		writePipe:     writer,
+		conn:          c.conn,
+		connWriteLock: &c.connWriteLock,
+		id:            id,
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+func (c *wsStreamCreator) Close() error {
+	c.closeAllStreamReaders(nil)
+	return nil
+}
+
+func (c *wsStreamCreator) CloseChan() <-chan bool {
+	return c.closeChan
+}
+
+func (c *wsStreamCreator) SetIdleTimeout(timeout time.Duration) {}
+
+func (c *wsStreamCreator) RemoveStreams(streams ...httpstream.Stream) {
+	for _, stream := range streams {
+		klog.V(4).Infof("RemoveStream: %d", stream.Identifier())
+		stream.Close()
+		c.removeStream(stream)
+	}
+}
+
+// ReadDemuxLoop is the reading processor for this endpoint of the websocket
+// connection. This loop reads the connection, and demultiplexes the data
+// into one of the individual stream pipes (by checking the stream id). This
+// loop can *not* be run concurrently, because there can only be one websocket
+// connection reader at a time (a read mutex would provide no benefit).
+func (c *wsStreamCreator) ReadDemuxLoop(bufferSize int, period time.Duration, deadline time.Duration) {
+	// Initialize and start the ping/pong heartbeat.
+	// h := newHeartbeat(c.conn, period, deadline)
+	// // Set initial timeout for websocket connection reading.
+	// if err := c.conn.SetReadDeadline(time.Now().Add(deadline)); err != nil {
+	// 	klog.Errorf("Websocket initial setting read deadline failed %v", err)
+	// 	return
+	// }
+	// go h.start()
+	// Buffer size must correspond to the same size allocated
+	// for the read buffer during websocket client creation. A
+	// difference can cause incomplete connection reads.
+	readBuffer := make([]byte, bufferSize)
+	for {
+		// NextReader() only returns data messages (BinaryMessage or Text
+		// Message). Even though this call will never return control frames
+		// such as ping, pong, or close, this call is necessary for these
+		// message types to be processed. There can only be one reader
+		// at a time, so this reader loop must *not* be run concurrently;
+		// there is no lock for reading. Calling "NextReader()" before the
+		// current reader has been processed will close the current reader.
+		// If the heartbeat read deadline times out, this "NextReader()" will
+		// return an i/o error, and error handling will clean up.
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if !ok || websocketErr.Code != gwebsocket.CloseNormalClosure {
+				c.closeAllStreamReaders(fmt.Errorf("next reader: %w", err))
+			}
+			return
+		}
+		// All remote command protocols send/receive only binary data messages.
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// It's ok to read just a single byte because the underlying library wraps the actual
```

Comment:
so we have a max of 254 / 2 ports, right?

Reply:
In order to provide the same functionality as SPDY port-forward, we will have to increase the stream id space (to approximately 2^31, I think). This will be one of three stream functionality items that we will have to implement over websockets (the other two being 1) a stream creation protocol, including headers sent from one endpoint to the other, and 2) the stream CLOSE signal sent to the other endpoint (like remotecommand). Marking resolved because of substantial changes.